crates/host/src/wasmbus/lattice.rs

```diff
@@ -0,0 +1,2954 @@
+                let mut heartbeat = Abortable::new(heartbeat, heartbeat_abort_reg);
+                heartbeat
+                    .by_ref()
+                    .for_each({
+                        let lattice = Arc::clone(&lattice);
+                        move |_| {
+                            let lattice = lattice.clone();
+                            async move {
+                                let heartbeat =
+                                    match serde_json::to_value(lattice.inventory().await) {
+                                        Ok(heartbeat) => heartbeat,
+                                        Err(e) => {
+                                            error!("failed to generate heartbeat: {e}");
+                                            return;
+                                        }
+                                    };
+
+                                if let Err(e) =
+                                    lattice.publish_event("host_heartbeat", heartbeat).await
+                                {
+                                    error!("failed to publish heartbeat: {e}");
+                                }
+                            }
+                        }
+                    })
+                    .await;
+                let deadline = { *lattice.stop_rx.borrow() };
+                lattice.stop_tx.send_replace(deadline);
+                if heartbeat.is_aborted() {
+                    info!(lattice = name, "heartbeat task gracefully stopped");
+                } else {
+                    error!(lattice = name, "heartbeat task unexpectedly stopped");
+                }
+            }
+        });
+
+        data.keys()
+            .await
+            .context("failed to read keys of lattice data bucket")?
+            .map_err(|e| anyhow!(e).context("failed to read lattice data stream"))
+            .try_filter_map(|key| async {
+                data.entry(key)
+                    .await
+                    .context("failed to get entry in lattice data bucket")
+            })
+            .for_each(|entry| async {
+                match entry {
+                    Ok(entry) => lattice.process_entry(entry, false).await,
+                    Err(err) => error!(%err, "failed to read entry from lattice data bucket"),
+                }
+            })
+            .await;
+
+        Ok(Arc::clone(&lattice))
+    }
+
+    async fn shutdown(&self) -> anyhow::Result<()> {
+        self.data_watch_abort.abort();
+        self.queue_abort.abort();
+        self.heartbeat_abort.abort();
+        self.policy_manager.policy_changes.abort();
+        // TODO add this back in
+        //let _ = try_join!(queue, data_watch).context("failed to await tasks")?;
+
+        self.rpc_nats
+            .flush()
+            .await
+            .context("failed to flush NATS connection")?;
+        // NOTE: Epoch interrupt thread will only stop once there are no more references to the engine
```

Comment:
I think this comment can be removed, and/or reconciled with #3330 which doesn't wait for the thread to close and rather closes the thread itself

Reply:
I rebased on #3330, but it wouldn't have picked up the comment. I'll remove it.